File: runtimes/python/src/twilic/wire.py
"""Low-level wire encoding and decoding primitives."""

from __future__ import annotations

import struct
from contextlib import contextmanager
from functools import wraps

from .errors import invalid_data, unexpected_eof, utf8_error


def encode_varuint(value: int, out: bytearray) -> None:
    if value < 0x80:
        out.append(value)
        return
    while True:
        b = value & 0x7F
        value >>= 7
        if value != 0:
            b |= 0x80
        out.append(b)
        if value == 0:
            break


def encode_zigzag(value: int) -> int:
    return (value << 1) ^ (value >> 63)


def decode_zigzag(value: int) -> int:
    return (value >> 1) ^ -(value & 1)


def encode_bytes(data: bytes, out: bytearray) -> None:
    encode_varuint(len(data), out)
    out.extend(data)


def encode_string(value: str, out: bytearray) -> None:
    encode_bytes(value.encode("utf-8"), out)


def encode_bitmap(bits: list[bool], out: bytearray) -> None:
    encode_varuint(len(bits), out)
    current = 0
    for i, bit in enumerate(bits):
        if bit:
            current |= 1 << (i % 8)
        if i % 8 == 7:
            out.append(current)
            current = 0
    if len(bits) % 8 != 0:
        out.append(current)


class Reader:
    __slots__ = ("_input", "_offset", "_depth", "_budget")

    def __init__(self, input_data: bytes) -> None:
        self._input = input_data
        self._offset = 0
        self._depth = 0
        self._budget = min(1 << 20, len(input_data) * 1024)

    def claim_output(self, count: int, width: int = 8) -> None:
        if count < 0 or count > 1 << 20:
            raise invalid_data("decode count limit exceeded")
        size = count * width
        if size > self._budget:
            raise invalid_data("decode output ratio exceeded")
        self._budget -= size

    def read_count(self, maximum: int = 1 << 20) -> int:
        count = self.read_varuint()
        if count > maximum:
            raise invalid_data("decode count limit exceeded")
        self.claim_output(count)
        return count

    @contextmanager
    def nested(self):
        if self._depth >= 64:
            raise invalid_data("decode depth limit exceeded")
        self._depth += 1
        try:
            yield
        finally:
            self._depth -= 1

    def position(self) -> int:
        return self._offset

    def is_eof(self) -> bool:
        return self._offset >= len(self._input)

    def read_u8(self) -> int:
        if self._offset >= len(self._input):
            raise unexpected_eof()
        b = self._input[self._offset]
        self._offset += 1
        return b

    def read_exact(self, n: int) -> bytes:
        end = self._offset + n
        if n < 0 or n > len(self._input) - self._offset:
            raise unexpected_eof()
        slice_ = self._input[self._offset : end]
        self._offset = end
        return slice_

    def read_varuint(self) -> int:
        shift = 0
        result = 0
        while True:
            if shift >= 64:
                raise invalid_data("varuint too large")
            b = self.read_u8()
            if shift == 63 and b & 0x7E:
                raise invalid_data("varuint too large")
            result |= (b & 0x7F) << shift
            if b & 0x80 == 0:
                return result
            shift += 7

    def read_i64_zigzag(self) -> int:
        encoded = self.read_varuint()
        return decode_zigzag(encoded)

    def read_bytes(self) -> bytes:
        n = self.read_varuint()
        return self.read_exact(n)

    def read_string(self) -> str:
        n = self.read_varuint()
        data = self.read_exact(n)
        try:
            return data.decode("utf-8")
        except UnicodeDecodeError as exc:
            raise utf8_error() from exc

    def read_bitmap(self) -> list[bool]:
        bit_count = self.read_count()
        byte_count = (bit_count + 7) // 8
        raw = self.read_exact(byte_count)
        bits = [False] * bit_count
        for i in range(bit_count):
            bits[i] = ((raw[i // 8] >> (i % 8)) & 1) == 1
        return bits


def new_reader(input_data: bytes) -> Reader:
    return Reader(input_data)


def read_u64_le(reader: Reader) -> int:
    b = reader.read_exact(8)
    return struct.unpack("<Q", b)[0]


def read_f64_le(reader: Reader) -> float:
    u = read_u64_le(reader)
    return struct.unpack("<d", struct.pack("<Q", u))[0]


def append_u64_le(out: bytearray, v: int) -> None:
    out.extend(struct.pack("<Q", v & 0xFFFFFFFFFFFFFFFF))


def append_f64_le(out: bytearray, v: float) -> None:
    append_u64_le(out, struct.unpack("<Q", struct.pack("<d", v))[0])


def bounded_decode(function):
    """Apply one shared recursion budget to nested decode operations."""

    @wraps(function)
    def wrapped(*args, **kwargs):
        reader = next(arg for arg in args if isinstance(arg, Reader))
        with reader.nested():
            return function(*args, **kwargs)

    return wrapped

File: runtimes/php/src/Twilic/Wire.php
<?php

declare(strict_types=1);

namespace Twilic;

final class Reader
{
    private string $input;
    private int $offset = 0;
    private int $depth = 0;
    private int $budget;

    public function __construct(string $inputData)
    {
        $this->input = $inputData;
        $this->budget = min(strlen($inputData), 1024) * 1024;
    }

    public function claimOutput(int $count): void
    {
        if ($count < 0 || $count > (1 << 20)) throw invalid_data('decode count limit exceeded');
        if ($count > intdiv($this->budget, 8)) throw invalid_data('decode output ratio exceeded');
        $this->budget -= $count * 8;
    }
    public function readCount(int $maximum = 1 << 20): int
    {
        $count = $this->readVaruint();
        if ($count < 0 || $count > $maximum) throw invalid_data('decode count limit exceeded');
        $this->claimOutput($count);
        return $count;
    }
    public function withDepth(callable $decode): mixed
    {
        if ($this->depth >= 64) throw invalid_data('decode depth limit exceeded');
        $this->depth++;
        try { return $decode(); } finally { $this->depth--; }
    }

    public function position(): int
    {
        return $this->offset;
    }

    public function isEof(): bool
    {
        return $this->offset >= strlen($this->input);
    }

    public function readU8(): int
    {
        if ($this->offset >= strlen($this->input)) {
            throw unexpected_eof();
        }
        $b = ord($this->input[$this->offset]);
        $this->offset++;
        return $b;
    }

    public function readExact(int $n): string
    {
        $end = $this->offset + $n;
        if ($n < 0 || $n > strlen($this->input) - $this->offset) {
            throw unexpected_eof();
        }
        $slice = substr($this->input, $this->offset, $n);
        $this->offset = $end;
        return $slice;
    }

    public function readVaruint(): int
    {
        $shift = 0;
        $result = 0;
        while (true) {
            if ($shift >= 64) {
                throw invalid_data('varuint too large');
            }
            $b = $this->readU8();
            if ($shift === 63 && ($b & 0x7E) !== 0) throw invalid_data("varuint too large");
            $result |= ($b & 0x7F) << $shift;
            if (($b & 0x80) === 0) {
                return $result;
            }
            $shift += 7;
        }
    }

    public function readI64Zigzag(): int
    {
        return decode_zigzag($this->readVaruint());
    }

    public function readBytes(): string
    {
        $n = $this->readVaruint();
        return $this->readExact($n);
    }

    public function readString(): string
    {
        $n = $this->readVaruint();
        $data = $this->readExact($n);
        if (!mb_check_encoding($data, 'UTF-8')) {
            throw utf8_error();
        }
        return $data;
    }

    /** @return list<bool> */
    public function readBitmap(): array
    {
        $bitCount = $this->readCount();
        $byteCount = intdiv($bitCount + 7, 8);
        $raw = $this->readExact($byteCount);
        $bits = array_fill(0, $bitCount, false);
        for ($i = 0; $i < $bitCount; $i++) {
            $bits[$i] = ((ord($raw[intdiv($i, 8)]) >> ($i % 8)) & 1) === 1;
        }
        return $bits;
    }
}

function encode_varuint(int $value, ByteBuffer $out): void
{
    if ($value < 0x80) {
        $out->append($value);
        return;
    }
    while (true) {
        $b = $value & 0x7F;
        $value >>= 7;
        if ($value !== 0) {
            $b |= 0x80;
        }
        $out->append($b);
        if ($value === 0) {
            break;
        }
    }
}

function encode_zigzag(int $value): int
{
    return ($value << 1) ^ ($value >> 63);
}

function decode_zigzag(int $value): int
{
    $u = ($value >> 1) ^ -($value & 1);
    if ($u > PHP_INT_MAX) {
        $u -= 0x10000000000000000;
    }
    return $u;
}

function encode_bytes(string $data, ByteBuffer $out): void
{
    encode_varuint(strlen($data), $out);
    $out->appendBytes($data);
}

function encode_string(string $value, ByteBuffer $out): void
{
    encode_bytes($value, $out);
}

/** @param list<bool> $bits */
function encode_bitmap(array $bits, ByteBuffer $out): void
{
    encode_varuint(count($bits), $out);
    $current = 0;
    foreach ($bits as $i => $bit) {
        if ($bit) {
            $current |= 1 << ($i % 8);
        }
        if ($i % 8 === 7) {
            $out->append($current);
            $current = 0;
        }
    }
    if (count($bits) % 8 !== 0) {
        $out->append($current);
    }
}

function new_reader(string $inputData): Reader
{
    return new Reader($inputData);
}

function read_u64_le(Reader $reader): int
{
    $b = $reader->readExact(8);
    $lo = unpack('V', substr($b, 0, 4))[1];
    $hi = unpack('V', substr($b, 4, 4))[1];
    return $lo + ($hi << 32);
}

function read_f64_le(Reader $reader): float
{
    $arr = unpack('e', $reader->readExact(8));
    return $arr[1];
}

function append_u64_le(ByteBuffer $out, int $v): void
{
    $out->appendBytes(pack('V', $v & 0xFFFFFFFF) . pack('V', ($v >> 32) & 0xFFFFFFFF));
}

function append_f64_le(ByteBuffer $out, float $v): void
{
    $out->appendBytes(pack('e', $v));
}
